Mathlib/LinearAlgebra/RootSystem/Basic.lean

```diff
@@ -67,50 +67,47 @@ open AddSubgroup (zmultiples)
 
 noncomputable section
 
-variable (ι : Type*) [Finite ι] {R M N : Type*}
+variable (ι R M N : Type*) [Finite ι]
   [CommRing R] [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N]
-  [IsReflexive R M] (e : N ≃ₗ[R] Dual R M)
 
 /-- Given two perfectly-paired `R`-modules `M` and `N`, a root pairing with finite indexing set `ι`
 is the data of an `ι`-indexed subset of `M` ("the roots") and an `ι`-indexed subset of `N`
 ("the coroots") satisfying the axioms familiar from the traditional theory of root systems / data.
 
 It exists to allow for a convenient unification of the theories of root systems and root data. -/
-structure RootPairing :=
+structure RootPairing extends PerfectPairing R M N :=
   root : ι ↪ M
   coroot : ι ↪ N
-  coroot_root_two : ∀ i, e (coroot i) (root i) = 2
+  coroot_root_two : ∀ i, toLin.flip (coroot i) (root i) = 2
   mapsTo_preReflection_root :
-    ∀ i, MapsTo (preReflection (root i) (e (coroot i))) (range root) (range root)
+    ∀ i, MapsTo (preReflection (root i) (toLin.flip (coroot i))) (range root) (range root)
   mapsTo_preReflection_coroot :
-    ∀ i, MapsTo (preReflection (coroot i) (e.flip (root i))) (range coroot) (range coroot)
+    ∀ i, MapsTo (preReflection (coroot i) (toLin (root i))) (range coroot) (range coroot)
 
 attribute [nolint docBlame] RootPairing.root
 attribute [nolint docBlame] RootPairing.coroot
 
 /-- A root datum is a root pairing for which the roots and coroots take values in finitely-generated
 free Abelian groups. -/
-abbrev RootDatum {X₁ X₂ : Type*}
+abbrev RootDatum (X₁ X₂ : Type*)
     [AddCommGroup X₁] [Free ℤ X₁] [Finite ℤ X₁]
-    [AddCommGroup X₂]
-    (e : X₂ ≃+ Dual ℤ X₁) :=
-  RootPairing ι e.toIntLinearEquiv
+    [AddCommGroup X₂] :=
+  RootPairing ι ℤ X₁ X₂
 
-variable (R M) in
 /-- A root system is a root pairing for which the roots span their ambient module and the perfect
 pairing is the canonical one between `M` and `Dual R M`. -/
-structure RootSystem extends RootPairing ι (LinearEquiv.refl R (Dual R M)) :=
+structure RootSystem extends RootPairing ι R M N :=
   span_eq_top : span R (range root) = ⊤
 
 namespace RootPairing
 
-variable {e ι}
-variable (P : RootPairing ι e) (i : ι)
+variable {ι R M N}
+variable (P : RootPairing ι R M N) (i : ι)
 
 /-- A root pairing is said to be crystallographic if the pairing between a root and coroot is
 always an integer.-/
 def IsCrystallographic : Prop :=
-  ∀ i, MapsTo (e (P.coroot i)) (range P.root) (zmultiples (1 : R))
+  ∀ i, MapsTo (P.toLin.flip (P.coroot i)) (range P.root) (zmultiples (1 : R))
```

Comment:
Do you think `P.toLin` could be handled by a coercion? If you define `P.flip` early on, then hopefully we also don't need `P.toLin.flip` here, and can just write `P.flip`.

Reply:
Yep, done.

Reply:
Oh wait, the point about the coercion is pending. Let me see.

Reply:
But in the previous design you could write `e x y`. I think it would be nice to be able to replicate that.

Reply:
But maybe your point is that this only makes sense if the pairing is an argument to `RootPairing` instead of a field of the structure.

Reply:
> Maybe your point is that this only makes sense if the pairing is an argument to `RootPairing` instead of a field of the structure.

Yes, it's partly this. I claim that if someone had `P : RootPairing ι R M N` and `m : M` then they would be surprised that `P m` made sense (and had type `Dual R N`).

I could unbundle the perfect pairing again and things would probably be fine. The key feature of the new design was the introduction of the type `PerfectPairing`; I included the bundling to avoid having issues with terms of type `RootPairing ι p.flip.flip`. Even though (thanks to the introduction of `PerfectPairing`) `p.flip.flip` is definitionally `p`, I could imagine this causing some pain (e.g., with rewrites).

Reply:
Incidentally another way to cut down on the explicit `P.toLin` and `P.toLin.flip` would be to introduce something like:
```lean
abbrev root' : Dual R N := P.toLin (P.root i)

abbrev coroot' : Dual R M := P.toLin.flip (P.coroot i)
```